extensions/pkg/controller/controlplane/reconciler.go

```diff
@@ -96,18 +101,27 @@ func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 		return reconcile.Result{}, err
 	}
 
-	if cp.DeletionTimestamp != nil {
+	operationType := gardencorev1beta1helper.ComputeOperationType(cp.ObjectMeta, cp.Status.LastOperation)
+
+	switch {
```

Comment:
It's now the third or fourth time that I see this switch which always does the same checks. Would it make sense to introduce a helper function that can be called by every controller and that returns the respective operation type? You could then call `r.migrate`, `r.delete`, etc., based on the returned operation type?

Reply:
I am not sure if I understand you correctly. The switch is not working with operationType, but a set of conditions. If we move this logic to а separate function that returns an operationType then another switch should be used in the consumer function.

Reply:
Yes, right, the idea was a separate function that gets the `extensionsv1alpha1.Object` as input and generate the operationType as output. You would still need this switch here, however, the "computations" for the various cases would be moved into this separate function and can be reused across all controllers. WDYT, would something like this make sense?

Reply:
I think this will lower the code duplication, but decrease simplicity and agility to change this switch in the future for a particular controller. Anyway what operation type should the function return if it is already migrated, I mean like in the first case? Also we still don't have lastOperationType=Restore, because it was introduced with the flow. Maybe do it like this now and change it after?

Reply:
Yes, sure, agreed, makes sense. Just wanted to throw it in so that we can give it a thought.

Reply:
We could do it with a factory which returns MigrateOperation, DeleteOperation, RestoreOperation, ReconcileOperation or NilOperation (for the first case) then for different controllers we could just extend the factory if changes are required. But I'm ok with doing this later on.